### PR TITLE
GH-43329: [C++] Order classes in flight/types.h according to Flight.proto

### PR DESCRIPTION
**Pull request by felipecrv**
### Rationale for this change

Identify the differences between Flight.proto and the hand-written types.h,.cc files.

### What changes are included in this PR?

 - Re-ordering of the classes

### Are these changes tested?

By existing tests.
* GitHub Issue: #43329

### CODE REVIEW

File: cpp/src/arrow/flight/serialization_internal.cc [hunk 1]
```diff
@@ -282,8 +283,8 @@ Status FromProto(const pb::BasicAuth& pb_basic_auth, BasicAuth* basic_auth) {
   return Status::OK();
 }
 
-Status FromProto(const pb::SchemaResult& pb_result, std::string* result) {
-  *result = pb_result.schema();
+Status FromProto(const pb::SchemaResult& pb_result, SchemaResult* result) {
```
Comment (felipecrv) [inline cpp/src/arrow/flight/serialization_internal.cc:286]: This is internal, so the API change is OK.

**Comment by felipecrv** [comment]
@lidavidm this might be annoying to review, but I split the changes in different commits to hopefully make it easier to understand what is going on.

**Comment by felipecrv** [comment]
I opened a PR focused on just fixing the CI issues (MSVC doesn't like the code): https://github.com/apache/arrow/pull/43332

Don't review this one just yet. I will rebase once the fix is merged.

File: cpp/src/arrow/flight/type_fwd.h [hunk 2]
```diff
@@ -55,11 +51,46 @@ class ServerCallContext;
...
-struct FlightData;
 class ServerTransport;
-}  // namespace internal
-}  // namespace flight
-}  // namespace arrow
+struct FlightData;
+}  // namespace arrow::flight::internal
+
+// Protobuf types from Flight.proto
+namespace arrow::flight::protocol {
```
Comment (lidavidm) [inline cpp/src/arrow/flight/type_fwd.h:64]: We don't want to expose these publicly.

File: cpp/src/arrow/flight/type_fwd.h [hunk 3]
```diff
@@ -55,11 +51,46 @@ class ServerCallContext;
...
+class GetSessionOptionsResult;
+class CloseSessionRequest;
+class CloseSessionResult;
+}  // namespace arrow::flight::protocol
+
+namespace arrow::internal {
+class Uri;
+}
+
+namespace google::protobuf {
```
Comment (lidavidm) [inline cpp/src/arrow/flight/type_fwd.h:94]: Same here. There shouldn't be any protobuf headers in our headers...

Comment (felipecrv) [reply cpp/src/arrow/flight/type_fwd.h:94]: Right. I will move these to an internal header.

Me next goal is being able to compile flight without linking to protobuf until the user actually links against `libarrow_flight` and the `libprotobuf` version of their choice.

**Comment by felipecrv** [comment]
I did the re-ordering from scratch. The whole PR is just ordering declarations and definitions now.

File: cpp/src/arrow/flight/types.cc [hunk 4]
```diff
@@ -151,18 +151,57 @@ Status MakeFlightError(FlightStatusCode code, std::string message,
...
-      return cmd == other.cmd;
-    default:
-      return false;
+  os << ']';
+
+  return os;
+}
+
+template <typename T>
+static std::ostream& operator<<(std::ostream& os, std::map<std::string, T> m) {
```
Comment (felipecrv) [inline cpp/src/arrow/flight/types.cc:167]: The `map` here needs to be taken by `const &`. I will do this in another PR.

File: cpp/src/arrow/flight/types.cc [hunk 5]
```diff
@@ -151,18 +151,57 @@ Status MakeFlightError(FlightStatusCode code, std::string message,
                        std::make_shared<FlightStatusDetail>(code, std::move(extra_info)));
 }
 
-bool FlightDescriptor::Equals(const FlightDescriptor& other) const {
-  if (type != other.type) {
-    return false;
+static std::ostream& operator<<(std::ostream& os, std::vector<std::string> values) {
```
Comment (felipecrv) [inline cpp/src/arrow/flight/types.cc:154]: The `vector` here needs to be taken by `const &`. I will do this in another PR. I just moved this function.

File: cpp/src/arrow/flight/types.h [hunk 6]
```diff
@@ -1078,6 +1068,29 @@ struct ARROW_FLIGHT_EXPORT CloseSessionRequest
...
+  // Subsequent requests with the same session produce a NOT_FOUND error.
+  kClosed,
+  // \brief The session close request is in progress.
+  //
+  // The client may retry the request.
+  kClosing,
+  // \brief The session is not closeable.
+  //
+  // The client should not retry the request.
+  kNotClosable
```
Comment (felipecrv) [inline cpp/src/arrow/flight/types.h:1089]: Funny typo here. It's "not_closEable" in `Flight.proto`.

**Review by lidavidm — APPROVED** [review]

**Comment by felipecrv** [comment]
CI failure is unrelated. So it's safe to merge.

**Pull request merged**